src/clr-angular/data/datagrid/providers/columns.service.ts

```diff
@@ -41,6 +41,37 @@ export class ColumnsService {
     return this.columnStates.filter(state => state.hideable).length > 0;
   }
 
+  get orderOfLastVisible(): number {
+    return Math.max(...this.columnStates.filter(state => !state.hidden).map(state => state.order));
+  }
+
+  get orderOfFirstVisible(): number {
+    return Math.min(...this.columnStates.filter(state => !state.hidden).map(state => state.order));
+  }
+
+  // First and last visible columns may change in 3 different cases:
+  // -  When columns are reordered
+  // -  When the current first/last gets hidden state
+  // -  When the column at first/last gets removed or a new column added at first/last index
+  private _isFirstVisibleChanged: Subject<void> = new Subject<void>();
+  private _isLastVisibleChanged: Subject<void> = new Subject<void>();
```

Comment:
nit: I would rather the first/last properties were kept together instead.

```
private _isFirstVisibleChanged: Subject<void> = new Subject<void>();
get checkFirstVisible(): Observable<void> {
    return this._isFirstVisibleChanged.asObservable();
}
isFirstVisibleChanged() {
    this._isFirstVisibleChanged.next();
}
// Then last visible
```

Reply:
Isn't it easier to look up properties when the same types are kept together? For me, at least, it's easier and faster to find properties when the same types are kept together.